### ISSUE TITLE
In some cases, `dune` needs to run cross-compiled executables

### ISSUE THREAD

**Issue by toots** [ocaml/dune#2571]
While working on https://github.com/toots/ocaml-sys-socket I noticed that, in order to have `ctypes` generate the proper `C` constants for the underlying socket structures when cross-compiling for windows, the binary that it generates needed to be run as a binary for the `windows` target.

I was able to work around it by using the `system` call instead of `run` and calling a wrapper that used `wine` or `wine64`. See: https://github.com/toots/ocaml-sys-socket/blob/master/src/sys-socket/types/dune#L10

I believe that this might become a fairly reasonable use case when writing low-level bindings. I know that `ocaml-ffmpeg` has the same issue.

Could it be possible to allow the user to run an executable as a target binary, possibly by also passing a wrapper for it, perhaps with default values such as `wine` for `windows`?

**Comment by avsm** [comment]
Interesting! How does this work if cross compiling from Linux? Do you have some `binfmt_misc` wrapper setup so that the `windows` target binary will work on Linux without the `wine` wrapper?

**Comment by toots** [comment]
Yes, I'm using `ocaml-config:system` as a script argument to discriminate:
* https://github.com/toots/ocaml-sys-socket/blob/1fbeaa6a47c4d4c871a3238d3af12ac01d284db1/src/sys-socket/types/dune#L15
* https://github.com/toots/ocaml-sys-socket/blob/master/src/sys-socket/generator/exec.sh

**Comment by avsm** [comment]
Thanks, this is also useful to guide the ctypes/dune builtin rules in the future as well.

**Comment by diml** [comment]
Would the method we use in configurator be enough for this as well? FTR, it works as follow: we compile a program where the constant we want to extract are surrounded by marker. Then we parse the produced binary, look for the markers and extract the value in between. This way we never need to execute the binary.

**Comment by toots** [comment]
@diml I'm not a specialist with windows exe format but if the constant is tucked in between markers in a static String, typically, I don't see why this wouldn't work. However, you'd still have to make sure that the binary is built for the correct target architecture.

@avsm Yes indeed! I've been meaning to write something about it. Pretty happy with the way the build system turned out with https://github.com/toots/ocaml-sys-socket it's both complex and neatly organized, it really did convince me of the power of dune, alone and in conjonction w/ ctypes.. We'll be looking at converting our bindings at savonet next.

**Comment by diml** [comment]
Agreed. It should be easy to copy&paste the method in ctypes BTW.

**Comment by avsm** [comment]
Very glad to hear all the pieces are fitting together, @toots!  The configurator method of examining the target binary is in Extract_obj at https://github.com/ocaml/dune/tree/master/src/configurator, and described here further https://dune.build/blog/configurator-constants/
